contracts/forwarder/Forwarder.sol

```diff
@@ -59,12 +59,25 @@ contract Forwarder is IForwarder {
         _verifyAndUpdateNonce(req);
 
         bytes memory callData = abi.encodePacked(req.data, req.from);
+        uint preBalance = address (this).balance;
+        require(preBalance >= req.value, "FWD: insufficient value");
```

Comment:
Why preBalance and not msg.value here?

Reply:
there are 2 ways to get value into the transaction:
1. sending eth to the forwarder just before calling execute (that's how a paymaster works)
2. passing "value" into execute()
checking `this.balance` uses the SUM of those two methods, so the rest of the code doesn't care what method was used to transfer in value.

Reply:
Why do we want to pass value before the `excute`?

Reply:
because a paymaster doesn't call `execute()`. Its the relayhub that calls it.
the the paymaster's preRelayedCall checks the `request.value` and transfer money into the forwarder, knowing the forwarder will use that money in the relayed call.
the "payable" mode is there for "completeness", but its unused by GSN.

Reply:
it will revert anyway on attempt to transfer more value than the current forwarder balance, why bother with this require?

Reply:
not strictly required. it just gives better reason, instead of empty string revert (we also add a revert string on our notion of "OOG")